### PR TITLE
[automatic failover] Implement HealtStatusManager + weighted endpoints

### PR DESCRIPTION
**Pull request by atakavci**
- weighted cluster seleciton
- Healtstatus manager with initial listener and registration logic
- pluggable health checker strategy  introduced,  these are draft NoOpStrategy, EchoStrategy, LagAwareStrategy,
- fix failing integration tests impacted by weighted clusters

### CODE REVIEW

File: src/main/java/redis/clients/jedis/HostAndPort.java [hunk 1]
```diff
@@ -2,7 +2,9 @@
 
 import java.io.Serializable;
 
-public class HostAndPort implements Serializable {
+import redis.clients.jedis.mcf.Endpoint;
+
+public class HostAndPort implements Serializable, Endpoint {
```
Comment (a-TODO-rov) [inline src/main/java/redis/clients/jedis/HostAndPort.java:7]: getHost() and getPort() methods are not marked as being overridden

File: src/main/java/redis/clients/jedis/UnifiedJedis.java [hunk 2]
```diff
@@ -238,19 +238,7 @@ public UnifiedJedis(ConnectionProvider provider, int maxAttempts, Duration maxTo
    */
   @Experimental
   public UnifiedJedis(MultiClusterPooledConnectionProvider provider) {
-    this(new CircuitBreakerCommandExecutor(provider, FailoverOptions.builder().build()), provider);
```
Comment (a-TODO-rov) [inline src/main/java/redis/clients/jedis/UnifiedJedis.java:241]: Unused import for `redis.clients.jedis.mcf.FailoverOptions`.

File: src/main/java/redis/clients/jedis/providers/MultiClusterPooledConnectionProvider.java [hunk 3]
```diff
@@ -240,8 +270,54 @@ public void validateTargetConnection(int multiClusterIndex) {
      * provided new index. Special care should be taken to confirm cluster/database availability AND potentially
      * cross-cluster replication BEFORE using this capability.
      */
-    public void setActiveMultiClusterIndex(int multiClusterIndex) {
+    // public void setActiveMultiClusterIndex(int multiClusterIndex) {
```
Comment (a-TODO-rov) [inline src/main/java/redis/clients/jedis/providers/MultiClusterPooledConnectionProvider.java]: Do we still need this commented code ?
Also, there is a link [https://github.com/redis/jedis/pull/4189/files#diff-51a7d130f973e51d707bbad52d03c1aaf39fcd7923a04c2f3593d545decbb02aR40](url) in the documentation, that prevents tests to run successfully.

File: src/main/java/redis/clients/jedis/providers/MultiClusterPooledConnectionProvider.java [hunk 4]
```diff
@@ -147,71 +148,100 @@ public MultiClusterPooledConnectionProvider(MultiClusterClientConfig multiCluste
...
-                lastClusterCircuitBreakerForcedOpen = true;
+        clusterWithHealthChange.setHealthStatus(newStatus);
 
-                throw new JedisConnectionException(
-                    "Cluster/database endpoint could not failover since the MultiClusterClientConfig was not "
-                        + "provided with an additional cluster/database endpoint according to its prioritized sequence. "
-                        + "If applicable, consider failing back OR restarting with an available cluster/database endpoint.");
-            } else activeMultiClusterIndex++;
+        if (newStatus.isHealthy()) {
+            if (clusterWithHealthChange.isFailbackEnabled() && activeCluster != clusterWithHealthChange) {
```
Comment (a-TODO-rov) [inline src/main/java/redis/clients/jedis/providers/MultiClusterPooledConnectionProvider.java:196]: You might try to skip the nested if.

File: src/main/java/redis/clients/jedis/providers/MultiClusterPooledConnectionProvider.java [hunk 5]
```diff
@@ -147,71 +148,100 @@ public MultiClusterPooledConnectionProvider(MultiClusterClientConfig multiCluste
...
+                // lets check if weighted switching is possible
+                Map.Entry<Endpoint, Cluster> failbackCluster = findWeightedFailbackCluster();
+                if (failbackCluster == clusterWithHealthChange
+                    && clusterWithHealthChange.getWeight() > activeCluster.getWeight()) {
+                    setActiveCluster(clusterWithHealthChange, false);
+                }
+            }
+        } else if (clusterWithHealthChange == activeCluster) {
```
Comment (a-TODO-rov) [inline src/main/java/redis/clients/jedis/providers/MultiClusterPooledConnectionProvider.java:204]: May be try to move as first condition with early return.

Comment (a-TODO-rov) [inline src/main/java/redis/clients/jedis/providers/MultiClusterPooledConnectionProvider.java:204]: `clusterWithHealthChange == activeCluster` or `clusterWithHealthChange.equals(activeCluster)`

File: src/main/java/redis/clients/jedis/providers/MultiClusterPooledConnectionProvider.java (hunk 4 above)
Comment (a-TODO-rov) [inline src/main/java/redis/clients/jedis/providers/MultiClusterPooledConnectionProvider.java:196]: `activeCluster != clusterWithHealthChange` or `!activeCluster.equals(clusterWithHealthChange)` ?

File: src/main/java/redis/clients/jedis/providers/MultiClusterPooledConnectionProvider.java [hunk 6]
```diff
@@ -147,71 +148,100 @@ public MultiClusterPooledConnectionProvider(MultiClusterClientConfig multiCluste
...
-            String originalClusterName = getClusterCircuitBreaker().getName();
+        Cluster clusterWithHealthChange = multiClusterMap.get(endpoint);
 
-            // Only increment if it can pass this validation otherwise we will need to check for NULL in the data path
-            if (activeMultiClusterIndex + 1 > multiClusterMap.size()) {
+        if (clusterWithHealthChange == null) return;
+        if (clusterWithHealthChange.isForcedUnhealthy()) return;
 
-                lastClusterCircuitBreakerForcedOpen = true;
+        clusterWithHealthChange.setHealthStatus(newStatus);
```
Comment (a-TODO-rov) [inline src/main/java/redis/clients/jedis/providers/MultiClusterPooledConnectionProvider.java:193]: What if `clusterWithHealthChange.equals(activeCluster)`?
Will another thread using the activeCluster at this moment be notified about the status switch ?

File: src/main/java/redis/clients/jedis/mcf/EchoStrategy.java [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+    }
+
+    @Override
+    public int getTimeout() {
+        return timeout;
+    }
+
+    @Override
+    public HealthStatus doHealthCheck(Endpoint endpoint) {
+        return "OK".equals(echo.apply(endpoint)) ? HealthStatus.HEALTHY : HealthStatus.UNHEALTHY;
```
Comment (vladvildanov) [inline src/main/java/redis/clients/jedis/mcf/EchoStrategy.java]: As I understand real echo execution will be added later, right?

File: src/main/java/redis/clients/jedis/mcf/HealthCheck.java [hunk 8]
```diff
@@ -0,0 +1,107 @@
...
+    public Endpoint getEndpoint() {
+        return endpoint;
+    }
+
+    public HealthStatus getStatus() {
+        return statusRef.get().getValue();
+    }
+
+    public void start() {
+        scheduler.scheduleAtFixedRate(this::healthCheck, 0, strategy.getInterval(), TimeUnit.MILLISECONDS);
```
Comment (vladvildanov) [inline src/main/java/redis/clients/jedis/mcf/HealthCheck.java:42]: Sometimes healthcheck isn't a recurring action (f.e initial healthcheck to define initial state of databases), It's nice to have an interface to just call it once

File: src/main/java/redis/clients/jedis/providers/MultiClusterPooledConnectionProvider.java (hunk 4 above)
Comment (atakavci) [reply src/main/java/redis/clients/jedis/providers/MultiClusterPooledConnectionProvider.java:196]: could you elaborate this?

Comment (atakavci) [reply src/main/java/redis/clients/jedis/providers/MultiClusterPooledConnectionProvider.java:196]: i think we should proceed only if its the same cluster instance in memory(since all the operations going on/through that particular one), not by using a some custom comparison, if we ever provide a custom comparison.

File: src/main/java/redis/clients/jedis/providers/MultiClusterPooledConnectionProvider.java (hunk 5 above)
Comment (atakavci) [reply src/main/java/redis/clients/jedis/providers/MultiClusterPooledConnectionProvider.java:204]: same as the one above

Comment (atakavci) [reply src/main/java/redis/clients/jedis/providers/MultiClusterPooledConnectionProvider.java:204]: it will require to check same condition multiple times, is that correct?

File: src/main/java/redis/clients/jedis/mcf/HealthCheck.java (hunk 8 above)
Comment (atakavci) [reply src/main/java/redis/clients/jedis/mcf/HealthCheck.java:42]: @vladvildanov is this referring to this statement?
"_The health of a database is either verified in the background (at a specific check interval) or when the circuit enters the half-open state._"

File: src/main/java/redis/clients/jedis/UnifiedJedis.java (hunk 2 above)
Comment (atakavci) [reply src/main/java/redis/clients/jedis/UnifiedJedis.java:241]: done

File: src/main/java/redis/clients/jedis/HostAndPort.java (hunk 1 above)
Comment (atakavci) [reply src/main/java/redis/clients/jedis/HostAndPort.java:7]: done

File: src/main/java/redis/clients/jedis/mcf/HealthCheck.java (hunk 8 above)
Comment (vladvildanov) [reply src/main/java/redis/clients/jedis/mcf/HealthCheck.java:42]: @atakavci it refers to this statement:


```The MultiDBClient establishes a connection to each of the databases. It performs an initial health-check to determine if the DB is considered healthy. ```

File: src/main/java/redis/clients/jedis/providers/MultiClusterPooledConnectionProvider.java (hunk 3 above)
Comment (atakavci) [reply src/main/java/redis/clients/jedis/providers/MultiClusterPooledConnectionProvider.java]: clean up done

File: src/main/java/redis/clients/jedis/mcf/EchoStrategy.java (hunk 7 above)
Comment (atakavci) [reply src/main/java/redis/clients/jedis/mcf/EchoStrategy.java]: correct! 
its almost ready to commit a change containing it. working on tests.

File: src/main/java/redis/clients/jedis/providers/MultiClusterPooledConnectionProvider.java (hunk 6 above)
Comment (atakavci) [reply src/main/java/redis/clients/jedis/providers/MultiClusterPooledConnectionProvider.java:193]: its just to flag the cluster with the new health status, setting status itself does not trigger or interrupt other threads. 
if the cluster receiving status is the active one and its not healthy anymore, then it needs to do a failover. 
To summarize the cases and related behavior for other threads/commands in case of failover;
- if command request shows up after switching to new cluster, it will be send to new cluster
- if command request sent but response not received, it will wait for a response/timeout (or we might go for fast termination of command, we are not there yet. TBH not even sure if there is a clean approach to do it, may be closing the related socket ?!)
- if command response received, then it basically means command completed
- if command request failed during request/response, then command would be retried on new cluster in respect to `retryOnFailover` setting

File: src/main/java/redis/clients/jedis/providers/MultiClusterPooledConnectionProvider.java (hunk 4 above)
Comment (a-TODO-rov) [reply src/main/java/redis/clients/jedis/providers/MultiClusterPooledConnectionProvider.java:196]: Yes. It's related to the comment bellow (on the else-if statement). If you move `else if (clusterWithHealthChange == activeCluster)` as first condition with early return than you will not need to check the second condition in `if (clusterWithHealthChange.isFailbackEnabled() && activeCluster != clusterWithHealthChange)`. And this way you will be able to remove the nested if statement by stacking the conditions like `if (newStatus.isHealthy() && clusterWithHealthChange.isFailbackEnabled())`.
Basically this suggestion is very optional and not a deal breaker, just a cleaner solution in my opinion.

File: src/main/java/redis/clients/jedis/mcf/EchoStrategy.java [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+        return interval;
+    }
+
+    @Override
+    public int getTimeout() {
+        return timeout;
+    }
+
+    @Override
+    public HealthStatus doHealthCheck(Endpoint endpoint) {
```
Comment (ggivo) [inline src/main/java/redis/clients/jedis/mcf/EchoStrategy.java:42]: Seems like the injected Endpoint is not useful, since in most cases you need additional configuration to perform a connection/call.

File: src/main/java/redis/clients/jedis/providers/MultiClusterPooledConnectionProvider.java (hunk 5 above)
Comment (a-TODO-rov) [reply src/main/java/redis/clients/jedis/providers/MultiClusterPooledConnectionProvider.java:204]: Yes. Also related to the comment above.

File: src/main/java/redis/clients/jedis/mcf/FailoverOptions.java [hunk 10]
```diff
@@ -8,9 +10,42 @@
 @Experimental
 public class FailoverOptions {
     private final boolean retryOnFailover;
+    private final StrategySupplier healthCheckStrategySupplier;
+    private final float weight;
```
Comment (ggivo) [inline src/main/java/redis/clients/jedis/mcf/FailoverOptions.java]: Shouldn't  weight  be a property of the ClusterConfig's itself, and not part of the FailoverOptions

File: src/main/java/redis/clients/jedis/mcf/FailoverOptions.java [hunk 11]
```diff
@@ -8,9 +10,42 @@
...
+    private boolean failback;
```
Comment (ggivo) [inline src/main/java/redis/clients/jedis/mcf/FailoverOptions.java]: Can we document the semantics of configuration options?
If I am getting it right, **failback**  means that this **Cluster** is **DISABLED** and should not be used even if `HEALTHY`?

File: src/main/java/redis/clients/jedis/providers/MultiClusterPooledConnectionProvider.java (hunk 6 above)
Comment (a-TODO-rov) [reply src/main/java/redis/clients/jedis/providers/MultiClusterPooledConnectionProvider.java:193]: Thanks for the summary.
I think i wasn't clear enough in this comment ...
I was referring to this change:
`private volatile Integer activeMultiClusterIndex = 1;` -> `private volatile Cluster activeCluster;`
So lets say you are switching the status here to the activeCluster -> when effectively this change will be visible for a concurrent thread ?
Basically the writes are synchronized with lock, but the reads are synchronized by the variable being declared volatile. But initially this variable was of a primitive kind and now is not. Having this in mind, should we expose a public setter for it at all ?
So to summarize all of the above my concern is - do you think we might be introducing some concurrent problem with the change, mentioned above ?

File: src/main/java/redis/clients/jedis/mcf/HealthStatusManager.java [hunk 12]
```diff
@@ -0,0 +1,79 @@
...
+
+    public void unregisterListener(Endpoint endpoint, HealthStatusListener listener) {
+        endpointListeners.computeIfPresent(endpoint, (k, v) -> {
+            v.remove(listener);
+            return v;
+        });
+        listeners.remove(listener);
+    }
+
+    public void notifyListeners(HealthStatusChangeEvent eventArgs) {
```
Comment (ggivo) [inline src/main/java/redis/clients/jedis/mcf/HealthStatusManager.java:33]: Since endpoint listeners are automatically also registered as global listeners then we will notify same listener twice.

Is this intentional?

File: src/main/java/redis/clients/jedis/mcf/FailoverOptions.java [hunk 13]
```diff
@@ -8,9 +10,42 @@
 @Experimental
 public class FailoverOptions {
     private final boolean retryOnFailover;
```
Comment (ggivo) [inline src/main/java/redis/clients/jedis/mcf/FailoverOptions.java]: My initial tought was that to be a Client wide configuration not Cluster wide.
If I remember correctly it's  meaning is if we want to retry failed inflight command during failover.

Do we want it configurable per cluster?

File: src/main/java/redis/clients/jedis/mcf/CircuitBreakerFailoverBase.java [hunk 14]
```diff
@@ -48,25 +48,29 @@ protected void clusterFailover(CircuitBreaker circuitBreaker) {
...
+            // if its the same one and there are no more clusters to failover to, 
+            // then throw an exception
+            else if (circuitBreaker == provider.getCluster().getCircuitBreaker() && !provider.canIterateOnceMore()) {
                 throw new JedisConnectionException(
                     "Cluster/database endpoint could not failover since the MultiClusterClientConfig was not "
                         + "provided with an additional cluster/database endpoint according to its prioritized sequence. "
                         + "If applicable, consider failing back OR restarting with an available cluster/database endpoint");
             }
+            // Ignore exceptions since we are already in a failure state
+        } catch (RuntimeException e) {
```
Comment (ggivo) [inline src/main/java/redis/clients/jedis/mcf/CircuitBreakerFailoverBase.java]: Whats' the idea to here?

File: src/main/java/redis/clients/jedis/mcf/EchoStrategy.java [hunk 15]
```diff
@@ -0,0 +1,50 @@
...
+
+    private int interval;
+    private int timeout;
+    private UnifiedJedis jedis;
+
+    public EchoStrategy(HostAndPort hostAndPort, JedisClientConfig jedisClientConfig) {
+       this(hostAndPort, jedisClientConfig, 1000, 1000);
+    }
+
+    public EchoStrategy(HostAndPort hostAndPort, JedisClientConfig jedisClientConfig, int interval, int timeout) {
```
Comment (ggivo) [inline src/main/java/redis/clients/jedis/mcf/EchoStrategy.java:20]: can we use some sort of Configration/Option  to wrap interval,timeout object so that in future we can extend with additinal oproperties without having to update the constructor

File: src/main/java/redis/clients/jedis/mcf/LagAwareStrategy.java [hunk 16]
```diff
@@ -0,0 +1,43 @@
...
+    }
+
+    @Override
+    public int getTimeout() {
+        return timeout;
+    }
+
+    @Override
+    public HealthStatus doHealthCheck(Endpoint endpoint) {
+        RedisRestAPIHelper helper = new RedisRestAPIHelper(endpoint.getHost(), String.valueOf(endpoint.getPort()),
```
Comment (ggivo) [inline src/main/java/redis/clients/jedis/mcf/LagAwareStrategy.java:28]: make sure we don't forget hardcoded values

File: src/main/java/redis/clients/jedis/mcf/RedisRestAPIHelper.java [hunk 17]
```diff
@@ -0,0 +1,95 @@
...
+ */
+class RedisRestAPIHelper {
+    // Required connection information - replace with your host, port, username, and
+    // password
+    private final String host;
+    private final String port;
+    private final String userName;
+    private final String password;
+
+    RedisRestAPIHelper(String host, String port, String userName, String password) {
```
Comment (ggivo) [inline src/main/java/redis/clients/jedis/mcf/RedisRestAPIHelper.java:26]: I would go with generic options class here, so that we can easily extend it, e.g tls, proxy ..

File: src/main/java/redis/clients/jedis/providers/MultiClusterPooledConnectionProvider.java [hunk 18]
```diff
@@ -130,7 +135,7 @@ public MultiClusterPooledConnectionProvider(MultiClusterClientConfig multiCluste
         for (ClusterConfig config : clusterConfigs) {
             GenericObjectPoolConfig<Connection> poolConfig = config.getConnectionPoolConfig();
 
-            String clusterId = "cluster:" + config.getPriority() + ":" + config.getHostAndPort();
+            String clusterId = "cluster:" + config.getFailoverOptions().getWeight() + ":" + config.getHostAndPort();
```
Comment (ggivo) [inline src/main/java/redis/clients/jedis/providers/MultiClusterPooledConnectionProvider.java]: Do we want to include the weight in cluster id?
In future they might become calculated dynamically at some point.

I was also under the impression that weight should be modifiable within the current design?

File: src/main/java/redis/clients/jedis/mcf/HealthStatus.java [hunk 19]
```diff
@@ -0,0 +1,19 @@
+package redis.clients.jedis.mcf;
+
+public enum HealthStatus {
+    HEALTHY(0x01), UNHEALTHY(0x02), FORCED_UNHEALTHY(0x04);
```
Comment (ggivo) [inline src/main/java/redis/clients/jedis/mcf/HealthStatus.java]: What is the intended use of FORCED_UNHEALTHY?

File: src/main/java/redis/clients/jedis/providers/MultiClusterPooledConnectionProvider.java (hunk 6 above)
Comment (atakavci) [reply src/main/java/redis/clients/jedis/providers/MultiClusterPooledConnectionProvider.java:193]: i dont think there is conccurrency issue since each time an event received, its only possible that related healtcheck thread is executing the `handleStatusChange`. So that thread is bound to send events on that individual cluster, at least the implementation does that with its current stage.

File: src/main/java/redis/clients/jedis/providers/MultiClusterPooledConnectionProvider.java [hunk 20]
```diff
@@ -147,71 +152,103 @@ public MultiClusterPooledConnectionProvider(MultiClusterClientConfig multiCluste
...
+                // lets check if weighted switching is possible
+                Map.Entry<Endpoint, Cluster> failbackCluster = findWeightedFailbackCluster();
+                if (failbackCluster == clusterWithHealthChange
+                    && clusterWithHealthChange.getWeight() > activeCluster.getWeight()) {
+                    setActiveCluster(clusterWithHealthChange, false);
+                }
+            }
+        } else if (clusterWithHealthChange == activeCluster) {
+            iterateActiveCluster();
+        }
```
Comment (atakavci) [inline src/main/java/redis/clients/jedis/providers/MultiClusterPooledConnectionProvider.java]: @a-TODO-rov, is this what you suggest? if not, could you provide code snippet on your suggestion? 

```suggestion
        if (clusterWithHealthChange == activeCluster && !newStatus.isHealthy()) {
            iterateActiveCluster();
            return;
        }
        if (newStatus.isHealthy() && clusterWithHealthChange.isFailbackEnabled() && activeCluster != clusterWithHealthChange) {
                // lets check if weighted switching is possible
                Map.Entry<Endpoint, Cluster> failbackCluster = findWeightedFailbackCluster();
                if (failbackCluster == clusterWithHealthChange
                    && clusterWithHealthChange.getWeight() > activeCluster.getWeight()) {
                    setActiveCluster(clusterWithHealthChange, false);
                }
            }
        } 
```

**Comment by dengliming** [comment]
不好Sincerely,  产品测试专用（勿动）
@atakavci commented on this pull request.



In src/main/java/redis/clients/jedis/providers/MultiClusterPooledConnectionProvider.java:
> +        if (newStatus.isHealthy()) {
+            if (clusterWithHealthChange.isFailbackEnabled() && activeCluster != clusterWithHealthChange) {
+                // lets check if weighted switching is possible
+                Map.Entry<Endpoint, Cluster> failbackCluster = findWeightedFailbackCluster();
+                if (failbackCluster == clusterWithHealthChange
+                    && clusterWithHealthChange.getWeight() > activeCluster.getWeight()) {
+                    setActiveCluster(clusterWithHealthChange, false);
+                }
+            }
+        } else if (clusterWithHealthChange == activeCluster) {
+            iterateActiveCluster();
+        }

@a-TODO-rov, is this what you suggest? if not, could you provide code snippet on your suggestion?
⬇️ Suggested change
-        if (newStatus.isHealthy()) {
-            if (clusterWithHealthChange.isFailbackEnabled() && activeCluster != clusterWithHealthChange) {
-                // lets check if weighted switching is possible
-                Map.Entry<Endpoint, Cluster> failbackCluster = findWeightedFailbackCluster();
-                if (failbackCluster == clusterWithHealthChange
-                    && clusterWithHealthChange.getWeight() > activeCluster.getWeight()) {
-                    setActiveCluster(clusterWithHealthChange, false);
-                }
-            }
-        } else if (clusterWithHealthChange == activeCluster) {
-            iterateActiveCluster();
-        }
+        if (clusterWithHealthChange == activeCluster && !newStatus.isHealthy()) {
+            iterateActiveCluster();
+            return;
+        }
+        if (newStatus.isHealthy() && clusterWithHealthChange.isFailbackEnabled() && activeCluster != clusterWithHealthChange) {
+                // lets check if weighted switching is possible
+                Map.Entry<Endpoint, Cluster> failbackCluster = findWeightedFailbackCluster();
+                if (failbackCluster == clusterWithHealthChange
+                    && clusterWithHealthChange.getWeight() > activeCluster.getWeight()) {
+                    setActiveCluster(clusterWithHealthChange, false);
+                }
+            }
+        } 



—Reply to this email directly, view it on GitHub, or unsubscribe.You are receiving this because you are subscribed to this thread.Message ID: ***@***.***>
[
{
***@***.***": "http://schema.org",
***@***.***": "EmailMessage",
"potentialAction": {
***@***.***": "ViewAction",
"target": "https://github.com/redis/jedis/pull/4189#pullrequestreview-3005372340",
"url": "https://github.com/redis/jedis/pull/4189#pullrequestreview-3005372340",
"name": "View Pull Request"
},
"description": "View this Pull Request on GitHub",
"publisher": {
***@***.***": "Organization",
"name": "GitHub",
"url": "https://github.com"
}
}
]

File: src/main/java/redis/clients/jedis/providers/MultiClusterPooledConnectionProvider.java [hunk 21]
```diff
@@ -250,79 +293,80 @@ public void setActiveMultiClusterIndex(int multiClusterIndex) {
...
-            if (activeMultiClusterIndex == multiClusterIndex)
+            if (activeCluster == cluster)
                 log.warn("Cluster/database endpoint '{}' successfully closed its circuit breaker", originalClusterName);
             else log.warn("Cluster/database endpoint successfully updated from '{}' to '{}'", originalClusterName,
-                getClusterCircuitBreaker(multiClusterIndex).getName());
-
-            activeMultiClusterIndex = multiClusterIndex;
-            lastClusterCircuitBreakerForcedOpen = false;
+                cluster.circuitBreaker.getName());
+            if (!activeCluster.isFailbackEnabled()) {
```
Comment (ggivo) [inline src/main/java/redis/clients/jedis/providers/MultiClusterPooledConnectionProvider.java:307]: I think we should keep the HealtStatus accurate based on actual health checks and keep the `failbaclEnabled` separate. Where needed, we can do  (cluster.isHealthy && cluster.isEnabled()

File: src/main/java/redis/clients/jedis/providers/MultiClusterPooledConnectionProvider.java [hunk 22]
```diff
@@ -147,71 +152,103 @@ public MultiClusterPooledConnectionProvider(MultiClusterClientConfig multiCluste
...
+        if (clusterWithHealthChange.isForcedUnhealthy()) return;
 
-            // Only increment if it can pass this validation otherwise we will need to check for NULL in the data path
-            if (activeMultiClusterIndex + 1 > multiClusterMap.size()) {
+        clusterWithHealthChange.setHealthStatus(newStatus);
 
-                lastClusterCircuitBreakerForcedOpen = true;
+        if (newStatus.isHealthy()) {
+            if (clusterWithHealthChange.isFailbackEnabled() && activeCluster != clusterWithHealthChange) {
+                // lets check if weighted switching is possible
```
Comment (ggivo) [inline src/main/java/redis/clients/jedis/providers/MultiClusterPooledConnectionProvider.java:197]: Looks like there is a concurency issue here.

If multiple Clusters change state to healthy, we can end up with `setActiveCluster()`  one that is  with the least weight,

File: src/main/java/redis/clients/jedis/providers/MultiClusterPooledConnectionProvider.java [hunk 23]
```diff
@@ -147,71 +152,103 @@ public MultiClusterPooledConnectionProvider(MultiClusterClientConfig multiCluste
...
+                Map.Entry<Endpoint, Cluster> failbackCluster = findWeightedFailbackCluster();
+                if (failbackCluster == clusterWithHealthChange
+                    && clusterWithHealthChange.getWeight() > activeCluster.getWeight()) {
+                    setActiveCluster(clusterWithHealthChange, false);
+                }
+            }
+        } else if (clusterWithHealthChange == activeCluster) {
+            iterateActiveCluster();
```
Comment (ggivo) [inline src/main/java/redis/clients/jedis/providers/MultiClusterPooledConnectionProvider.java]: Should this be syncronised with CB failover?
If CB try to failover, at the same time HealtCheck triggers handleStatusChange
we can end up invoking iterateActiveCluster twice.

In general I think is good idea if we can manage the failover in single place.
Currently if failover is initiated by CB transition to OPEN state, failover is triggered  in the CircuitBreakerCommandExecutor (with proper locking), if it is triggred by handleStatusChange,  iterateActiveCluster() is invoked there and this I think can lead to syncronisation problems.

**Comment by atakavci** [comment]
hey @dengliming , 
what about these comments you had via email. Could you explain and if this is something not planned, please remove or disable the automation or whatever process this is. 
This is kind of spamming our PR's , so thank you for your cooperation.

**Comment by dengliming** [comment]
Not goodSincerely,  产品测试专用（勿动）atakavci left a comment (redis/jedis#4189)
hey @dengliming ,
what about these comments you had via email. Could you explain and if this is something not planned, please remove or disable the automation or whatever process this is.
This is kind of spamming our PR's , so thank you for your cooperation.

—Reply to this email directly, view it on GitHub, or unsubscribe.You are receiving this because you were mentioned.Message ID: ***@***.***>

[
{
***@***.***": "http://schema.org",
***@***.***": "EmailMessage",
"potentialAction": {
***@***.***": "ViewAction",
"target": "https://github.com/redis/jedis/pull/4189#issuecomment-3061286205",
"url": "https://github.com/redis/jedis/pull/4189#issuecomment-3061286205",
"name": "View Pull Request"
},
"description": "View this Pull Request on GitHub",
"publisher": {
***@***.***": "Organization",
"name": "GitHub",
"url": "https://github.com"
}
}
]

File: src/main/java/redis/clients/jedis/mcf/EchoStrategy.java (hunk 9 above)
Comment (atakavci) [reply src/main/java/redis/clients/jedis/mcf/EchoStrategy.java:42]: this method signature is from [HealthCheckStrategy](https://github.com/redis/jedis/pull/4189/files#diff-410dd555814c16bbb7ee407e03065e8764a89de2ca5cee065c510ebe9360fae9) interface, where it is responsible for identifiying related Endpoint. Any other information is not considered to be relevant, at least at this level. 
The configuration can be passed through the `StrategySupplier`  in 'FailoverOptions` where both `endpoint` and `clientConfig` is provided to supplier.
Design is flexible and covers almost all possible use cases, which might be over-flexible. Also brings some complexity as we discussed.

File: src/main/java/redis/clients/jedis/mcf/FailoverOptions.java (hunk 10 above)
Comment (atakavci) [reply src/main/java/redis/clients/jedis/mcf/FailoverOptions.java]: i ll check if i can move ;
- `weight` and `healthCheckStrategySupplier` to `ClusterConfig`
- `retryOnFailover` and `failback` to `MultiClusterClientConfig`

File: src/main/java/redis/clients/jedis/mcf/FailoverOptions.java (hunk 11 above)
Comment (atakavci) [reply src/main/java/redis/clients/jedis/mcf/FailoverOptions.java]: thanks for pointing to this;
i think this needs to be revised anyway so i will get help from AI to add javadoc especially for these unclear definitions later stages. Let's agree on rules and expectations on failback behaviour first.

Comment (atakavci) [reply src/main/java/redis/clients/jedis/mcf/FailoverOptions.java]: to answer your second Q;
this is intended for enabling/disabling "failing back to this cluster" once its failed over from.

File: src/main/java/redis/clients/jedis/mcf/HealthStatusManager.java (hunk 12 above)
Comment (atakavci) [reply src/main/java/redis/clients/jedis/mcf/HealthStatusManager.java:33]: good catch, it is a mistake for sure.

File: src/main/java/redis/clients/jedis/mcf/FailoverOptions.java (hunk 13 above)
Comment (atakavci) [reply src/main/java/redis/clients/jedis/mcf/FailoverOptions.java]: i ll check if i can move ;
- `weight` and `healthCheckStrategySupplier` to `ClusterConfig`
- `retryOnFailover` and `failback` to `MultiClusterClientConfig`

File: src/main/java/redis/clients/jedis/mcf/CircuitBreakerFailoverBase.java (hunk 14 above)
Comment (atakavci) [reply src/main/java/redis/clients/jedis/mcf/CircuitBreakerFailoverBase.java]: its left over from my initial experiments on circuit breaker behaviour..
clean up done.

File: src/main/java/redis/clients/jedis/mcf/EchoStrategy.java (hunk 15 above)
Comment (atakavci) [reply src/main/java/redis/clients/jedis/mcf/EchoStrategy.java:20]: not sure if this deserves a new type for option, let me follow and add the type if we need another field/parameter.

File: src/main/java/redis/clients/jedis/mcf/LagAwareStrategy.java (hunk 16 above)
Comment (atakavci) [reply src/main/java/redis/clients/jedis/mcf/LagAwareStrategy.java:28]: FWIW, `LagAwareStrategy` is nothing more than an idea as public API at this stage;  sorry that i couldn't properly filter my changes/commits to this PR. And its bigger than i initially anticipated.
I take this this in to its own small PR.

File: src/main/java/redis/clients/jedis/mcf/RedisRestAPIHelper.java (hunk 17 above)
Comment (atakavci) [reply src/main/java/redis/clients/jedis/mcf/RedisRestAPIHelper.java:26]: same comment as in the one for `LagAwareStrategy`

File: src/main/java/redis/clients/jedis/providers/MultiClusterPooledConnectionProvider.java (hunk 18 above)
Comment (atakavci) [reply src/main/java/redis/clients/jedis/providers/MultiClusterPooledConnectionProvider.java]: Absolutly right, TBH i dont even know what this value is doing there.

File: src/main/java/redis/clients/jedis/mcf/HealthStatus.java (hunk 19 above)
Comment (atakavci) [reply src/main/java/redis/clients/jedis/mcf/HealthStatus.java]: i ll do the change to introduce an aggregated status function(to gather the values from CB, HealtChecks and cluster enabled/disables) as we agreed in our dedicated session.

File: src/main/java/redis/clients/jedis/providers/MultiClusterPooledConnectionProvider.java (hunk 21 above)
Comment (atakavci) [reply src/main/java/redis/clients/jedis/providers/MultiClusterPooledConnectionProvider.java:307]: i ll go for an aggregated status function but still need to get my head around for this to be optimal and final(may be not final but mature). 
Want to avoid tracking multiple states here and there, as much as possible. 
To me, this kind of setup is a prime candidate for turning into status hell.

File: src/main/java/redis/clients/jedis/providers/MultiClusterPooledConnectionProvider.java (hunk 22 above)
Comment (atakavci) [reply src/main/java/redis/clients/jedis/providers/MultiClusterPooledConnectionProvider.java:197]: there is no race condition here since both `findWeightedFailbackCluster` and `setActiveCluster` is thread safe, where former is inherently safe while latter is safe by use of a `ReentrantLock`.
to explain better, if we assume two healthy cluster shows up, what is important here is sequence; 
if the less weighted shows up early, even with involving concurrency constructs, there is no escape from switching to it since we never know higher weighted would show up.

File: src/main/java/redis/clients/jedis/providers/MultiClusterPooledConnectionProvider.java (hunk 4 above)
Comment (atakavci) [reply src/main/java/redis/clients/jedis/providers/MultiClusterPooledConnectionProvider.java:196]: @a-TODO-rov, provided a code block [here](https://github.com/redis/jedis/pull/4189#discussion_r2197528925) to verify we are on same track, please check and let me know.

File: src/main/java/redis/clients/jedis/providers/MultiClusterPooledConnectionProvider.java (hunk 5 above)
Comment (atakavci) [reply src/main/java/redis/clients/jedis/providers/MultiClusterPooledConnectionProvider.java:204]: @a-TODO-rov, provided a code block [here](https://github.com/redis/jedis/pull/4189#discussion_r2197528925) to verify we are on same track, please check and let me know.

File: src/main/java/redis/clients/jedis/providers/MultiClusterPooledConnectionProvider.java (hunk 23 above)
Comment (atakavci) [reply src/main/java/redis/clients/jedis/providers/MultiClusterPooledConnectionProvider.java]: if they both trigger `iterateActiveCluster` they will  end up with same highest weighted healthy cluster in the iterate process anyway, 
which would do no-change in theory.
But i admit i am not %100 comfortable with how transitions are managed and agree to build the logic and safety into single place. 
i will save time to investigate and build corner case test for this.

**Comment by dengliming** [comment]
nubi 要吃牛肉，真是伤脑筋qianming

**Review by ggivo — APPROVED** [review]
LGTM

Let's start merging them in the feature branch so we have a consistent state there
